onnxruntime/core/session/node_attributes_abi_impl.cc

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+#include "core/common/status.h"
+#include "core/framework/error_code_helper.h"
+#include "core/session/ort_apis.h"
+#include "core/graph/basic_types.h"
+#include "core/framework/data_types.h"
+
+ORT_API_STATUS_IMPL(OrtApis::CreateNodeAttributes, _Outptr_ OrtNodeAttributes** attributes_) {
+  onnxruntime::NodeAttributes* attributes = new onnxruntime::NodeAttributes;
+  *attributes_ = reinterpret_cast<OrtNodeAttributes*>(attributes);
+  return onnxruntime::ToOrtStatus(onnxruntime::Status::OK());
+}
+
+#define ATTR_SETTER_IMPL(ctype, apiName, enumType, field)                                                  \
+  ORT_API_STATUS_IMPL(OrtApis::NodeAttributes_Set_##apiName,                                               \
+                      _Inout_ OrtNodeAttributes* attributes_,                                              \
+                      _In_ const char* name,                                                               \
+                      ctype value) {                                                                       \
+    API_IMPL_BEGIN                                                                                         \
+    ORT_ENFORCE(attributes_, "attributes must be non-null.");                                              \
+    onnxruntime::NodeAttributes* attributes = reinterpret_cast<onnxruntime::NodeAttributes*>(attributes_); \
+    ONNX_NAMESPACE::AttributeProto attribute_proto;                                                        \
+    attribute_proto.set_name(name);                                                                        \
+    attribute_proto.set_type(enumType);                                                                    \
+    attribute_proto.set_##field(value);                                                                    \
+    attributes->insert(std::make_pair(name, attribute_proto));                                             \
```

Comment:
>std::make_pair(name, attribute_proto) [](http://example.com/codeflow?start=23&length=37)

Can you use emplace and move attribute into the map instead of copying?